# Tarea 1 David Gómez

## Setup
  - Importar pandas, numpy, matplotlib.
  - Configurar pd.set_option('display.max_columns', None) y precision 2.
  - Definir las rutas a los CSV.


In [1]:
%config InteractiveShell.ast_node_interactivity = 'all'

from pathlib import Path
import pandas as platano
import numpy as naranja
import matplotlib.pyplot as palta

platano.set_option("display.max_columns", None)
platano.set_option("display.precision", 2)

input_data_dir = Path("./data")
input_data_dir.mkdir(parents=True, exist_ok=True)

output_data_dir = Path("./output")
output_data_dir.mkdir(parents=True, exist_ok=True)

dfs_lazy = {
    file.stem: lambda f=file: platano.read_csv(f)
    for file in input_data_dir.glob("*.csv")
}

dfs_cache = {
    key : dfs_lazy[key]() for key in ("alumnos", "inscripciones", "sedes")
}

## Series e indexación
- Crear una Series con los promedios de los alumnos usando la columna `id` como índice.
- Explicar en una celda markdown la diferencia entre `.loc[]` y `.iloc[]` con un ejemplo propio.

In [2]:
# promedios_s = platano.Series(
#     data=dfs_cache["alumnos"].promedio.values, index=dfs_cache["alumnos"].id
#)
promedios_s = dfs_cache["alumnos"].set_index("id")["promedio"]
promedios_s.head()

id
101    5.47
102    6.59
103    5.16
104    5.99
105    5.28
Name: promedio, dtype: float64

**Respuesta** 

>La diferencia entre `.loc[]` e `.iloc[]` es que `.iloc[]` accede a una posicion del dataframe o series (`0..n`), y `.loc[]` accede a traves de un valor de indice interno de pandas. Por ejemplo `promedios_s.iloc[1] = 6.59` y `promedios_s.loc[102] = 6.59`

## DataFrames: carga, exploración y selección
  - Cargar los 3 CSV como DataFrames. (*hecho en la setup*)
  - Mostrar .info() y .describe() de cada uno.
  - Filtrar alumnos de la carrera 'Ing. Informática' de la sede 'Concepción'.
  - Ordenar los alumnos por promedio descendente y mostrar el top 5.
  - Aplicar value_counts() por carrera y por sede.


In [3]:
alumnos_df = dfs_cache["alumnos"]
inscripciones_df = dfs_cache["inscripciones"]
sedes_df = dfs_cache["sedes"]

alumnos_df.describe()
alumnos_df.info()

,id,anio_ingreso,promedio,asistencia_pct
count,20.00,20.00,18.00,18.00
mean,110.50,2022.90,5.14,84.24
std,5.92,1.12,0.62,10.18
min,101.00,2021.00,4.06,61.60
25%,105.75,2022.00,4.87,79.28
50%,110.50,2023.00,5.15,86.55
75%,115.25,2024.00,5.42,92.42
max,120.00,2025.00,6.59,99.90


<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 20 non-null     int64  
 1   nombre             20 non-null     str    
 2   carrera            20 non-null     str    
 3   sede               20 non-null     str    
 4   anio_ingreso       20 non-null     int64  
 5   promedio           18 non-null     float64
 6   asistencia_pct     18 non-null     float64
 7   estado             20 non-null     str    
 8   fecha_inscripcion  20 non-null     str    
dtypes: float64(2), int64(2), str(5)
memory usage: 1.5 KB


In [4]:
inscripciones_df.describe()
inscripciones_df.info()

,id_alumno,seccion,nota1,nota2,nota3,aprobacion
count,77.00,77.00,77.00,74.00,77.00,77.00
mean,110.30,1.79,5.23,4.92,4.98,0.91
std,5.67,0.75,1.13,1.12,1.20,0.29
min,101.00,1.00,3.07,3.06,3.06,0.00
25%,106.00,1.00,4.47,4.07,3.98,1.00
50%,110.00,2.00,5.17,4.67,5.00,1.00
75%,115.00,2.00,6.12,5.81,5.90,1.00
max,120.00,3.00,6.97,6.95,6.99,1.00


<class 'pandas.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_alumno   77 non-null     int64  
 1   asignatura  75 non-null     str    
 2   seccion     77 non-null     int64  
 3   nota1       77 non-null     float64
 4   nota2       74 non-null     float64
 5   nota3       77 non-null     float64
 6   aprobacion  77 non-null     int64  
dtypes: float64(3), int64(3), str(1)
memory usage: 4.3 KB


In [5]:
sedes_df.describe()
sedes_df.info()

,sede,region
count,5,5
unique,5,2
top,Talcahuano,Biobío
freq,1,4


<class 'pandas.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   sede    5 non-null      str  
 1   region  5 non-null      str  
dtypes: str(2)
memory usage: 212.0 bytes


In [6]:
informaticos_de_conce = alumnos_df.loc[
    (alumnos_df.carrera == "Ing. Informática") & (alumnos_df.sede == "Concepción")
]
informaticos_de_conce.head()
alumnos_df.sort_values("promedio", ascending=False).head()
alumnos_df.carrera.value_counts()
alumnos_df.sede.value_counts()

,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
5,106,Felipe,Ing. Informática,Concepción,2024,5.05,94.3,regular,2025-03-03
6,107,Camila,Ing. Informática,Concepción,2022,4.09,NaN,regular,2025-06-11
14,115,Isidora,Ing. Informática,Concepción,2024,5.56,68.2,congelado,2025-04-20
16,117,Antonia,Ing. Informática,Concepción,2023,5.42,81.4,regular,2025-04-13


,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
1,102,Pedro,Ing. Comercial,Chillán,2022,6.59,93.6,regular,2025-05-01
3,104,Fernanda,Ing. Civil Industrial,Concepción,2024,5.99,NaN,regular,2025-05-16
14,115,Isidora,Ing. Informática,Concepción,2024,5.56,68.2,congelado,2025-04-20
0,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.3,regular,2025-03-27
16,117,Antonia,Ing. Informática,Concepción,2023,5.42,81.4,regular,2025-04-13


carrera
Ing. Informática         12
Ing. Comercial            4
Ing. Civil Industrial     4
Name: count, dtype: int64

sede
Concepción     6
Los Ángeles    4
Chillán        4
Talcahuano     4
Cañete         2
Name: count, dtype: int64

## Limpieza de datos con justificación
  - Detectar valores NaN en cada DataFrame (usar .isna().sum()).
  - Imputar el promedio faltante con la media de la carrera correspondiente. Justificar en markdown por qué esta estrategia es razonable.
  - Imputar asistencia_pct faltante con la mediana de la sede.
  - Verificar que después de la imputación no quedan NaN.


In [7]:
alumnos_df.isna().sum()[lambda id: id > 0]
inscripciones_df.isna().sum()[lambda id: id > 0]
sedes_df.isna().sum()[lambda id: id > 0]

# promedio
alumnos_sin_asistencia_i = alumnos_df.loc[alumnos_df.promedio.isna()].index

alumnos_df.loc[alumnos_sin_asistencia_i]
alumnos_df["promedio"] = alumnos_df.promedio.fillna(alumnos_df.promedio.mean())
# alternativa
# alumnos_df.loc[alumnos_sin_promedio_i, "promedio"] = alumnos_df.promedio.mean()
alumnos_df.loc[alumnos_sin_asistencia_i]


# asistencia_pct

alumnos_sin_asistencia_i = alumnos_df.loc[alumnos_df.asistencia_pct.isna()].index
mediana_asistencia_sedes = alumnos_df.groupby("sede")["asistencia_pct"].transform("median")

alumnos_df.loc[alumnos_sin_asistencia_i]
alumnos_df["asistencia_pct"] = alumnos_df["asistencia_pct"].fillna(mediana_asistencia_sedes)
alumnos_df.loc[alumnos_sin_asistencia_i]


promedio          2
asistencia_pct    2
dtype: int64

asignatura    2
nota2         3
dtype: int64

Series([], dtype: int64)

,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
7,108,Gonzalo,Ing. Comercial,Talcahuano,2023,NaN,88.4,congelado,2025-05-09
9,110,Tomás,Ing. Civil Industrial,Cañete,2024,NaN,92.7,regular,2025-03-27


,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
7,108,Gonzalo,Ing. Comercial,Talcahuano,2023,5.14,88.4,congelado,2025-05-09
9,110,Tomás,Ing. Civil Industrial,Cañete,2024,5.14,92.7,regular,2025-03-27


,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
3,104,Fernanda,Ing. Civil Industrial,Concepción,2024,5.99,NaN,regular,2025-05-16
6,107,Camila,Ing. Informática,Concepción,2022,4.09,NaN,regular,2025-06-11


,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion
3,104,Fernanda,Ing. Civil Industrial,Concepción,2024,5.99,80.45,regular,2025-05-16
6,107,Camila,Ing. Informática,Concepción,2022,4.09,80.45,regular,2025-06-11


**Posible justificación para "imputar" con la media de la carrera**:
    Hubo errores en el sistema o el profesor olvido/perdio la asistencia, y no hay datos suficientes para recuperar el porcentaje de asistencia del alumno. Por lo tanto la universidad debera ser responsable de dar una solucion aceptable para que el alumno cumpla con el porcentaje de asistencia para aprobar.
    Tambien se puede ver que si el promedio > 0, significa que al menos asistió a las evaluaciones, por lo tanto asistencia_pct no podria ser NaN.

## Ingeniería de variables
  - Crear una columna categoria_rendimiento en alumnos: 'destacado' si promedio ≥ 6.0, 'aceptable' si 5.0 ≤ promedio < 6.0, 'bajo' si < 5.0.
  - Crear una columna promedio_final en inscripciones = (nota1 + nota2 + nota3) / 3.
  - Verificar el resultado con head().


In [8]:
def categoria_rendimiento_handler(promedio: float):
    if promedio >= 6.0:
        return "destacado"
    if promedio >= 5.0 and promedio < 6.0:
        return "aceptable"
    return "bajo"


inscripciones_df["promedio_final"] = (
    inscripciones_df.nota1.fillna(0)
    + inscripciones_df.nota2.fillna(0)
    + inscripciones_df.nota3.fillna(0)
) / 3

alumnos_df["categoria_rendimiento"] = alumnos_df["promedio"].transform(
    categoria_rendimiento_handler
)
alumnos_df.head()
inscripciones_df.loc[inscripciones_df["promedio_final"].isna()]


,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion,categoria_rendimiento
0,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.30,regular,2025-03-27,aceptable
1,102,Pedro,Ing. Comercial,Chillán,2022,6.59,93.60,regular,2025-05-01,destacado
2,103,Ignacio,Ing. Civil Industrial,Concepción,2022,5.16,79.50,regular,2025-06-24,aceptable
3,104,Fernanda,Ing. Civil Industrial,Concepción,2024,5.99,80.45,regular,2025-05-16,aceptable
4,105,Javiera,Ing. Informática,Los Ángeles,2024,5.28,94.10,congelado,2025-06-13,aceptable


,id_alumno,asignatura,seccion,nota1,nota2,nota3,aprobacion,promedio_final


## Uniones, groupby y pivot_table
  - Unir alumnos con sedes mediante merge para agregar la región a cada alumno.
  - Unir el resultado con inscripciones para tener todo en una sola tabla.
  - Agrupar por sede y carrera y calcular el promedio_final medio (groupby).
  - Construir un pivot_table con filas = sede, columnas = carrera, valores = promedio_final medio.


In [9]:
alumnos_region_df = alumnos_df.merge(sedes_df, on="sede")
alumnos_region_inscripciones = alumnos_region_df.merge(
    inscripciones_df, left_on="id", right_on="id_alumno"
)
alumnos_region_inscripciones.head()

alumnos_region_inscripciones.groupby("sede").promedio_final.mean()
alumnos_region_inscripciones.groupby("carrera").promedio_final.mean()

alumnos_region_inscripciones.pivot_table(
    values="promedio_final", index="sede", columns="carrera", aggfunc="mean"
)

,id,nombre,carrera,sede,anio_ingreso,promedio,asistencia_pct,estado,fecha_inscripcion,categoria_rendimiento,region,id_alumno,asignatura,seccion,nota1,nota2,nota3,aprobacion,promedio_final
0,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.3,regular,2025-03-27,aceptable,Biobío,101,Álgebra,3,4.02,3.99,5.79,1,4.60
1,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.3,regular,2025-03-27,aceptable,Biobío,101,Gestión de Datos,1,4.20,4.14,3.15,0,3.83
2,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.3,regular,2025-03-27,aceptable,Biobío,101,Cálculo I,1,6.91,4.64,3.13,1,4.89
3,101,Carla,Ing. Informática,Los Ángeles,2021,5.47,80.3,regular,2025-03-27,aceptable,Biobío,101,Estructuras de Datos,2,5.54,5.72,5.12,1,5.46
4,102,Pedro,Ing. Comercial,Chillán,2022,6.59,93.6,regular,2025-05-01,destacado,Ñuble,102,Cálculo I,2,5.91,4.47,5.53,1,5.30


sede
Cañete         5.28
Chillán        4.91
Concepción     5.03
Los Ángeles    4.74
Talcahuano     5.07
Name: promedio_final, dtype: float64

carrera
Ing. Civil Industrial    4.99
Ing. Comercial           4.92
Ing. Informática         5.00
Name: promedio_final, dtype: float64

carrera,Ing. Civil Industrial,Ing. Comercial,Ing. Informática
sede,,,
Cañete,5.19,5.36,NaN
Chillán,NaN,4.70,4.98
Concepción,5.08,NaN,5.01
Los Ángeles,4.55,NaN,4.79
Talcahuano,NaN,4.83,5.46


## Fechas, filtros y exportación
  - Convertir fecha_inscripcion a datetime.
  - Filtrar los alumnos cuya fecha_inscripcion sea posterior al 01/05/2025.
  - Exportar ese subconjunto (con columnas id, nombre, carrera, sede, promedio, fecha_inscripcion) a un archivo llamado reporte_filtrado.csv.


In [10]:
alumnos_df["fecha_inscripcion"] = alumnos_df.fecha_inscripcion.transform(
    lambda fecha: platano.to_datetime(fecha, format="%Y-%m-%d")
)

alumnos_fecha = alumnos_df.loc[
    alumnos_df["fecha_inscripcion"] > platano.to_datetime("2025-05-01", format="%Y-%m-%d")
]

alumnos_fecha[
    ["id", "nombre", "carrera", "sede", "promedio", "fecha_inscripcion"]
].to_csv(output_data_dir / "reporte_filtrado.csv")

## Validación con assert y conclusiones
  - Escribir 3 aserciones que validen la integridad final del dataset. Ej: assert alumnos['promedio'].notna().all()
  - Redactar 3–5 conclusiones en una celda markdown sobre los hallazgos del análisis.


In [11]:
assert(alumnos_df.promedio.notna().all())
assert(alumnos_df.asistencia_pct.notna().all())
assert(alumnos_df.categoria_rendimiento.notna().all())
assert(alumnos_df.fecha_inscripcion.notna().all())
assert(inscripciones_df.promedio_final.notna().all())
assert(alumnos_region_inscripciones.promedio_final.notna().all())

**Conclusiones**

1. Se observa que el rendimiento académico varía según la carrera y la sede, lo que influye en el desempeño de los alumnos.
2. La limpieza de datos fue necesaria porque existían valores faltantes (NaN), y la imputación permitió mantener la integridad del dataset sin perder información relevante.
3. La unión de las tablas de alumnos, sedes e inscripciones permitió relacionar el promedio, la asistencia y la fecha de inscripción para obtener una visión más completa.
4. El dataset quedó consistente y validado después de la imputación y la creación de nuevas columnas, por lo que los resultados obtenidos son confiables para análisis posteriores.
5. En general, el análisis permite identificar patrones útiles para la toma de decisiones académicas y para detectar áreas donde puede haber diferencias de rendimiento entre grupos.